Public/Src/Engine/Dll/Distribution/Grpc/ServerInterceptor.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using BuildXL.Engine.Distribution.Grpc;
+using BuildXL.Engine.Tracing;
+using BuildXL.Utilities.Instrumentation.Common;
+using Grpc.Core;
+using Grpc.Core.Interceptors;
+
+namespace BuildXL.Engine.Distribution
+{
+    internal class ServerInterceptor : Interceptor
+    {
+        private readonly LoggingContext m_loggingContext;
+        private readonly string m_buildId;
+
+        public ServerInterceptor(LoggingContext loggingContext, string buildId)
+        {
+            m_loggingContext = loggingContext;
+            m_buildId = buildId;
+        }
+
+        private void InterceptCallContext(ServerCallContext context)
+        {
+            string sender = context.Host;
+            string traceId = string.Empty;
+            string method = context.Method;
+            string senderBuildId = string.Empty;
+
+            foreach (var kvp in context.RequestHeaders)
+            {
+                if (kvp.Key == GrpcConstants.TraceIdKey)
+                {
+                    traceId = new Guid(kvp.ValueBytes).ToString();
+                }
+                else if (kvp.Key == GrpcConstants.BuildIdKey)
+                {
+                    senderBuildId = kvp.Value;
+                }
+            }
+
+            if (!string.IsNullOrEmpty(senderBuildId) && senderBuildId != m_buildId)
+            {
+                throw new RpcException(

```

Comment:
>RpcException [](start = 26, length = 12)

Is throwing an exception, how this should be done? Did you verify that the exception makes it back to the client? Often in cache code, we just get a somewhat unhelpful message about exception being logged on server without text indicating the details of what happened.

Reply:
Discussed in-person. Basically, when we throw an exception on the server side, a failed message is sent to the sender with the exception details

---
In reply to: [272744985](https://github.com/Microsoft/BuildXL/pull/91#discussion_r272744985) [](ancestors = 272744985)